ddtrace/propagation/http.py

```diff
@@ -123,28 +92,49 @@ def my_controller(url, headers):
             return Context()
 
         try:
-            trace_id = HTTPPropagator.extract_trace_id(headers)
-            parent_span_id = HTTPPropagator.extract_parent_span_id(headers)
-            sampling_priority = HTTPPropagator.extract_sampling_priority(headers)
-            origin = HTTPPropagator.extract_origin(headers)
-
-            if sampling_priority is not None:
-                sampling_priority = int(sampling_priority)
-
-            return Context(
-                trace_id=trace_id,
-                span_id=parent_span_id,
-                sampling_priority=sampling_priority,
-                dd_origin=origin,
+            normalized_headers = {name.lower(): v for name, v in headers.items()}
+
+            trace_id = HTTPPropagator._extract_header_value(
+                POSSIBLE_HTTP_HEADER_TRACE_IDS,
+                normalized_headers,
+                default=0,
             )
-        # If headers are invalid and cannot be parsed, return a new context and log the issue.
-        except Exception:
-            log.debug(
-                "invalid x-datadog-* headers, trace-id: %s, parent-id: %s, priority: %s, origin: %s",
-                headers.get(HTTP_HEADER_TRACE_ID, 0),
-                headers.get(HTTP_HEADER_PARENT_ID, 0),
-                headers.get(HTTP_HEADER_SAMPLING_PRIORITY),
-                headers.get(HTTP_HEADER_ORIGIN, ""),
-                exc_info=True,
+            parent_span_id = HTTPPropagator._extract_header_value(
+                POSSIBLE_HTTP_HEADER_PARENT_IDS,
+                normalized_headers,
+                default=0,
+            )
+            sampling_priority = HTTPPropagator._extract_header_value(
+                POSSIBLE_HTTP_HEADER_SAMPLING_PRIORITIES,
+                normalized_headers,
+            )
+            origin = HTTPPropagator._extract_header_value(
+                POSSIBLE_HTTP_HEADER_ORIGIN,
+                normalized_headers,
             )
+
+            # Try to parse values into their expected types
+            try:
+                if sampling_priority is not None:
+                    sampling_priority = int(sampling_priority)
+
+                return Context(
+                    # DEV: Do not allow `0` for trace id or span id, use None instead
+                    trace_id=int(trace_id) or None,
+                    span_id=int(parent_span_id) or None,
+                    sampling_priority=sampling_priority,
+                    dd_origin=origin,
+                )
+            # If headers are invalid and cannot be parsed, return a new context and log the issue.
+            except (TypeError, ValueError):
+                log.debug(
```

Comment:
It'd be good to have tests ensuring that debug logs are generated, but I think that's not required for the core change here. I'll make note of it in our backlog.